### PR TITLE
leafletDraw

### PR DESCRIPTION
**Pull request by nmccready**
Don't merge this yet. It is mainly to start getting eyes on this.

Still needs:

- more specs
- examples
- documentation

Can make an additional PR as this is 37 files already.

### CODE REVIEW

File: .editorconfig [hunk 1]
```diff
@@ -9,7 +9,7 @@ root = true
 
 # Change these settings to your own preference
 indent_style = space
-indent_size = 4
+indent_size = 2
```
Comment (nmccready) [inline .editorconfig]: Start using 2 space indent. I'll fix it as I go; or I can do all the files in another follow up PR.

File: grunt/aliases.yaml [hunk 2]
```diff
@@ -4,10 +4,7 @@ test:
  - 'test-e2e'
 
 test-unit:
- - 'karma:unit'
```
Comment (nmccready) [inline grunt/aliases.yaml]: got rid of grunt-karma

File: grunt/babel.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
+'use strict';
```
Comment (nmccready) [inline grunt/babel.js]: we can now use es6

File: grunt/utils/karma.js [hunk 4]
```diff
@@ -0,0 +1,33 @@
+const karmaLib = require('karma');
```
Comment (nmccready) [inline grunt/utils/karma.js]: Karma runner, and this works quite well with gulp with little modification.

File: src/directives/draw.js [hunk 5]
```diff
@@ -0,0 +1,99 @@
+(() => {
+  const directiveName = 'lfDraw';
+
+  angular.module('ui-leaflet')
+  .config(function($provide) {
+    /*
+    Extend root leaflet directive,
+    */
+    return $provide.decorator('leafletDirective', function($delegate) {
+      var directive;
```
Comment (nmccready) [inline src/directives/draw.js]: extends leaflet.js , which should be done this way for all other directives

File: src/directives/draw.js [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+          var options;
+          if (!_deferred || _deferred.resolvedDefer) {
+            _deferred = $q.defer();
+          }
+          if (_optionsEditedInDirective) {
+            return;
+          }
+          options = leafletScope[directiveName] || {};
+          if (isDefined(options.control)) {
+            options.control.promised(_deferred.promise);
```
Comment (nmccready) [inline src/directives/draw.js]: To do this can go away once I make leafletData and directiveControls extendable.

File: test/karma-unit.conf.js [hunk 7]
```diff
@@ -1,80 +0,0 @@
...
-            'bower_components/leaflet.markercluster/dist/leaflet.markercluster.js',
-            'bower_components/leaflet.vector-markers/dist/Leaflet.vector-markers.js',
-            mainLib,
-            'test/unit/bootstrap.coffee',
-            'test/unit/*.js',
-            'test/unit/**/*.js',
-            'test/unit/**/*.coffee',
-            'bower_components/Leaflet.PolylineDecorator/leaflet.polylineDecorator.js',
-            //do not include those specs for jasmine html runner by karma kama_jasmine_runner.html
-            {pattern:'**/**/**/*.coffee', included: false},
```
Comment (nmccready) [inline test/karma-unit.conf.js]: This glob is what was crashing karma 0.13 and making so slow. It makes all files get read in like node_modules, and bower_comonents. Basically memory blows up. This is now fixed.

File: test/karma.conf.js [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+      'dist/src/**/*.js',
+      //tests
+      'test/unit/bootstrap.coffee',
+      'test/unit/*.js',
+      'test/unit/**/*.js',
+      'test/unit/**/*.coffee',
+      'bower_components/Leaflet.PolylineDecorator/leaflet.polylineDecorator.js',
+      {
+        pattern: 'test/**/*.coffee',
+        included: false
```
Comment (nmccready) [inline test/karma.conf.js]: consume less ignores

File: test/unit/bootstrap.coffee [hunk 9]
```diff
@@ -1,21 +1,25 @@
 window.ngLeafLetTestGlobals = {}
 
+afterEach ->
+  window.doTestLog = false
```
Comment (nmccready) [inline test/unit/bootstrap.coffee]: Lighten up the logging. Only turn doTestLog explicitly if you need to debug logs in your spec.

File: test/unit/services/leafletHelpersSpec.coffee [hunk 10]
```diff
@@ -1,53 +1,53 @@
...
+  beforeEach ->
+    module('ui-leaflet')
+    inject (_$compile_, _$rootScope_, _leafletData_, _leafletHelpers_) ->
+      @$compile = _$compile_
+      @$rootScope = _$rootScope_
+      @leafletData = _leafletData_
+      @subject = _leafletHelpers_
+
+  describe 'isTruthy', ->
     beforeEach ->
```
Comment (nmccready) [inline test/unit/services/leafletHelpersSpec.coffee]: indenting only

**Comment by nmccready** [comment]
@jessertaylor @elesdoar @simison

File: .editorconfig (hunk 1 above)
Comment (simison) [inline .editorconfig]: Editors tend to pick what's around, so it might be good idea to do one massive PR instead of leaving mixed styles behind.

Comment (nmccready) [inline .editorconfig]: Id rather not do all of it right now as it makes it even harder to review. Ill do it immediately after this PR.

Comment (simison) [inline .editorconfig]: Yeah, absolutely, indentation PR has to have only spaces changed and nothing else.

We had this with Meanjs and it was good to watch for a moment with minimum amount of PR's pending.

**Comment by elesdoar** [comment]
should this PR be on leaflet-1.X branch? We talk about plugins for version 2, no?

**Comment by elesdoar** [comment]
On the another hand, I like this changes, :+1:

File: src/services/events/leafletDrawEvents.js [hunk 11]
```diff
@@ -0,0 +1,26 @@
+angular.module('ui-leaflet')
+.factory('leafletDrawEvents', function(leafletEventsHelpersFactory) {
+
+  class DrawEvents extends leafletEventsHelpersFactory{
```
Comment (nmccready) [inline src/services/events/leafletDrawEvents.js]: BTW I thought the Babel compiled code to this was quite comical. As people who hate CoffeeScript babel's output is way way harder to read. I almost would rather revert this to ES5.

Comment (nmccready) [inline src/services/events/leafletDrawEvents.js]: https://github.com/realtymaps/ui-leaflet/blob/dev/nmccready/issue_105_leaflet_draw/dist/ui-leaflet.js#L5011-L5039

Comment (nmccready) [inline src/services/events/leafletDrawEvents.js]: vs
coffee```
angular.module('ui-leaflet')
.factory 'leafletDrawEvents', (leafletEventsHelpersFactory) ->

  class DrawEvents extends leafletEventsHelpersFactory
    constructor: () ->
      super('leafletDirectiveDraw', 'draw')
    

    getAvailableEvents: () ->
      return [
        'created'
        'edited'
        'deleted'
        'drawstart'
        'drawstop'
        'editstart'
        'editstop'
        'deletestart'
        'deletestop'].map (n) -> 
          return 'draw:' + n

    new DrawEvents();
```

yeilds:
```
var extend = function(child, parent) { for (var key in parent) { if (hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
  hasProp = {}.hasOwnProperty;

angular.module('ui-leaflet').factory('leafletDrawEvents', function(leafletEventsHelpersFactory) {
  var DrawEvents;
  return DrawEvents = (function(superClass) {
    extend(DrawEvents, superClass);

    function DrawEvents() {
      DrawEvents.__super__.constructor.call(this, 'leafletDirectiveDraw', 'draw');
    }

    DrawEvents.prototype.getAvailableEvents = function() {
      return ['created', 'edited', 'deleted', 'drawstart', 'drawstop', 'editstart', 'editstop', 'deletestart', 'deletestop'].map(function(n) {
        return 'draw:' + n;
      });
    };

    new DrawEvents();

    return DrawEvents;

  })(leafletEventsHelpersFactory);
});

// ---
// generated by coffee-script 1.9.2
```

**Comment by nmccready** [comment]
@elesdoar no, I need these changes now for work. Plus this is backwards compatible but done in a more modular way. I plan on switching the other directives to be like this as well with the angular.config.

File: package.json [hunk 12]
```diff
@@ -9,6 +9,11 @@
     "type": "git",
     "url": "https://github.com/angular-ui/ui-leaflet"
   },
+  "scripts": {
+    "test": "grunt travis --verbose",
+    "karma": "karma start ./test/karma.conf.js"
```
Comment (nmccready) [inline package.json]: Note besides `grunt karma` you can run `npm run karma` now.

**Comment by nmccready** [comment]
I plan on having a follow up PR for documentation and more specs. As well as fixing directiveControls, and leafletData.

**Comment by nmccready** [comment]
@elesdoar your right, I will put all the leaflet draw stuff into a sep repo. This PR will become just karma, grunt, and babel fixes.

**Comment by nmccready** [comment]
Closing for plugin / component repo for ui-leaflet-draw.

https://github.com/angular-ui/ui-leaflet-draw

**Pull request closed without merging**